File: java/leetcode/backtracking/sudoku_solver.java
/**
 * 37. Sudoku Solver
 * Write a program to solve a Sudoku puzzle by filling the empty cells.

A sudoku solution must satisfy all of the following rules:

Each of the digits 1-9 must occur exactly once in each row.
Each of the digits 1-9 must occur exactly once in each column.
Each of the the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes of the grid.
Empty cells are indicated by the character '.'.
Note:

The given board contain only digits 1-9 and the character '.'.
You may assume that the given Sudoku puzzle will have a single unique solution.
The given board size is always 9x9.
 */

public class sodoku_solver {

    private boolean[][] rowsUsed = new boolean[9][10];
    private boolean[][] colsUsed = new boolean[9][10];
    private boolean[][] cubesUsed = new boolean[9][10];
    private char[][] board;

    public void solveSudoku(char[][] board) {
        this.board = board;
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                int num = board[i][j] - '0';
                rowsUsed[i][num] = true;
                colsUsed[j][num] = true;
                cubesUsed[cubeNum(i,j)][num] = true;
            }
        }
        backtracking(0,0);
    }

    private boolean backtracking(int row, int col) {
        while (row < 9 && board[row][col] != '.') {
            row = col == 8 ? row + 1 : row;
            col = col == 8 ? 0 : col + 1;
        }
        if (row == 9) {
            return true;
        }
        for (int num = 1; num <= 9; num++) {
            if (rowsUsed[row][num] || colsUsed[col][num] || cubesUsed[cubeNum(row, col)][num]) {
                continue;
            }
            rowsUsed[row][num] = colsUsed[col][num] = cubesUsed[cubeNum(row, col)][num] = true;
            board[row][col] = (char)(num + '0');
            if (backtracking(row, col)) {
                return true;
            }
            board[row][col] = '.';
            rowsUsed[row][num] = colsUsed[col][num] = cubesUsed[cubeNum(row,col)][num] = false;
        }
        return false;
    }

    private int cubeNum(int i, int j) {
        int r = i / 3;
        int c = j / 3;
        return r * 3 + c;
    }
}

File: golang/leetcode/backtracking/sudoku_solver.go
package backtracking

// 37. Sudoku Solver
var (
	rowsUsed  [][]bool
	colsUsed  [][]bool
	cubesUsed [][]bool
)

func solveSudoku(board [][]byte) {
	rowsUsed, colsUsed, cubesUsed = make([][]bool, 9), make([][]bool, 9), make([][]bool, 9)
	for i := range rowsUsed {
		rowsUsed[i], colsUsed[i], cubesUsed[i] = make([]bool, 10), make([]bool, 10), make([]bool, 10)
	}

	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			if board[i][j] == '.' {
				continue
			}
			num := int(board[i][j] - '0')
			rowsUsed[i][num] = true
			colsUsed[j][num] = true
			cubesUsed[cubeNum(i, j)][num] = true
		}
	}

	sudokuTrack(0, 0, board)
}

func sudokuTrack(row int, col int, board [][]byte) bool {
	for row < 9 && board[row][col] != '.' {
		if col == 8 {
			row = row + 1
			col = 0
		} else {
			col = col + 1
		}
	}
	if row == 9 {
		return true
	}

	for num := 1; num <= 9; num++ {
		if rowsUsed[row][num] || colsUsed[col][num] || cubesUsed[cubeNum(row, col)][num] {
			continue
		}
		rowsUsed[row][num], colsUsed[col][num], cubesUsed[cubeNum(row, col)][num] = true, true, true
		board[row][col] = byte(num + '0')
		if sudokuTrack(row, col, board) {
			return true
		}
		board[row][col] = '.'
		rowsUsed[row][num], colsUsed[col][num], cubesUsed[cubeNum(row, col)][num] = false, false, false
	}
	return false
}

func cubeNum(i, j int) int {
	r, c := i/3, j/3
	return r*3 + c
}
